File: irrgarten/src/irrgarten/UI/TextUI.java

package irrgarten.UI;

import irrgarten.Directions;
import irrgarten.GameState;
import java.util.Scanner;


public class TextUI {
    
    private static Scanner in = new Scanner(System.in);
    
    private char readChar() {
        String s = in.nextLine();     
        return s.charAt(0);
    }
    

    public Directions nextMove() {
        System.out.print("Where? ");
        
        Directions direction = Directions.DOWN;
        boolean gotInput = false;
        
        while (!gotInput) {
            char c = readChar();
            switch(c) {
                case 'w':
                    System.out.print(" UP\n");
                    direction = Directions.UP;
                    gotInput = true;
                    break;
                case 's':
                    System.out.print(" DOWN\n");
                    direction = Directions.DOWN;
                    gotInput = true;
                    break;
                case 'd':
                    System.out.print("RIGHT\n");
                    direction = Directions.RIGHT;
                    gotInput = true;
                    break;
                case 'a':
                    System.out.print(" LEFT\n");
                    direction = Directions.LEFT;
                    gotInput = true;    
                    break;
            }
        }    
        return direction;
    }
    
    public void showGame(GameState gameState) {
        System.out.println("\t\t CURRENT GAMESTATE \n");
        System.out.println("\t\tMONSTERS\n" + gameState.getMonsters() + "\n");
        System.out.println("\t\tLABYRINTHV\n" + gameState.getLabyrinthv() + "\n");
        System.out.println("\t\tPLAYERS\n" + gameState.getPlayers());
        System.out.println("CURRENT PLAYER: " + (char) gameState.getCurrentPlayer());
        System.out.println("IS THERE A WINNER?: " + gameState.isWinner() + "\n");
        System.out.println("\t\tLOG\n" + gameState.getLog() + "\n");
    }
    
}

File: irrgarten_ruby/textUI.rb
# frozen_string_literal: true

require 'io/console'
require_relative 'directions'
require_relative 'game_state'

module UI
  class TextUI
    # https://gist.github.com/acook/4190379
    def read_char
      $stdin.echo = false
      $stdin.raw!

      input = $stdin.getch
      if input == "\e"
        begin
          input << $stdin.read_nonblock(3)
        rescue StandardError
          nil
        end
        begin
          input << $stdin.read_nonblock(2)
        rescue StandardError
          nil
        end
      end
    ensure
      $stdin.echo = true
      $stdin.cooked!

      return input
    end

    def next_move
      print 'Where? '
      got_input = false
      until got_input
        c = read_char
        case c
        when 'w'
          puts 'UP ARROW'
          output = Irrgarten::Directions::UP
          got_input = true
        when 's'
          puts 'DOWN ARROW'
          output = Irrgarten::Directions::DOWN
          got_input = true
        when 'd'
          puts 'RIGHT ARROW'
          output = Irrgarten::Directions::RIGHT
          got_input = true
        when 'a'
          puts 'LEFT ARROW'
          output = Irrgarten::Directions::LEFT
          got_input = true
        when "\u0003"
          puts 'CONTROL-C'
          got_input = true
          exit(1)
        else
          # error
        end
      end
      output
    end

    def show_game(game_state)
      puts "CURRENT GAME STATE\n"
      puts "MONSTERS:\n #{game_state.monsters}\n"
      puts "PLAYERS\n:#{game_state.players}\n"
      puts "LABYRINTH:\n#{game_state.labyrinth}\n"
      puts "CURRENT PLAYER: #{game_state.current_player}\n"
      puts "IS THERE A WINNER?: #{game_state.winner}\n"
      puts "LOG:\n#{game_state.log}\n"
    end
  end
end
